exchange/bitswap/bitswap.go

```diff
@@ -375,6 +373,10 @@ func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg
 			k := b.Cid()
 			log.Event(ctx, "Bitswap.GetBlockRequest.End", k)
 
+			for _, ses := range bs.SessionsForBlock(k) {
+				ses.receiveBlockFrom(p, b)
```

Comment:
That race condition exists in the code today, technically, if you get the timing just right, a call to get could return not found from the blockstore, then the entire put could happen, then the bitswap getblock could be called. 

Adding a notifications framework to the blockstore feels like overkill, and the only other real solutions involve locking, or an extra call to `Has` (which to be fair, can be easily cached).

At one point we did call blockstore.Get on each block again inside bitswap, but i remember it being removed at some point in the name of datastore performance.